notecard/main.go

```diff
@@ -1,107 +1,109 @@
-		panic(err)
+	if req.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		log.Printf("%s: Method not allowed", w)
+		return
 	}
 
-	// Print the resulting JSON string
-	fmt.Println(string(jsonString))
-
-	note_rsp, err := card.Transaction(data)
+	body, err := io.ReadAll(req.Body)
 	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		panic(err)
+		handleError(w, err, "while reading request body")
+		return
 	}
+	req.Body.Close()
 
-	note_rsp_json, err := json.Marshal(note_rsp)
+	note_rsp, err := s.card.TransactionJSON(body)
 	if err != nil {
-		fmt.Println("Error encoding JSON:", err)
-		w.WriteHeader(http.StatusInternalServerError)
-		panic(err)
+		handleError(w, err, "while performing a card transaction")
+		return
 	}
+	log.Printf("notecard response: %s", note_rsp)
 
 	// Set the Content-Type header to application/json
 	w.Header().Set("Content-Type", "application/json")
-	w.Write(note_rsp_json)
-
-	w.WriteHeader(http.StatusOK)
+	_, err = w.Write(note_rsp)
+	if err != nil {
+		log.Printf("error writing response: %v", err)
+		return
+	}
 }
 
-func setupNotecard(protocol Transport) *notecard.Context {
-	var card *notecard.Context
-	var err error
+func setupNotecard(protocol string) (*notecard.Context, error) {
+	log.Printf("Setting up Notecard with protocol: %s\n", protocol)
 
-	fmt.Printf("Setting up Notecard with protocol: %s\n", protocol)
+	if protocol != Serial && protocol != I2C {
+		return nil, fmt.Errorf("unsupported transport protocol: %v", protocol)
+	}
 
-	if protocol == Transport("serial") {
-		card, err = notecard.OpenSerial("/dev/tty.usbmodemNOTE1", 9600)
-		if err != nil {
-			fmt.Printf("Error opening Notecard: %v\n", err)
-			panic(err)
-		}
-	} else if protocol == Transport("i2c") {
-		card, err = notecard.OpenI2C("/dev/i2c-1", 0x17)
+	if protocol == Serial {
+		card, err := notecard.OpenSerial("/dev/tty.usbmodemNOTE1", 9600)
 		if err != nil {
-			fmt.Printf("Error opening Notecard: %v\n", err)
-			panic(err)
+			return nil, fmt.Errorf("error opening Notecard: %v", err)
 		}
-	} else {
-		fmt.Printf("Missing transport protocol\n")
+		return card, nil
+	}
+
+	card, err := notecard.OpenI2C("", 0x17)
+	if err != nil {
+		return nil, fmt.Errorf("error opening Notecard: %v", err)
+	}
+
+	status := map[string]interface{}{
+		"req": "card.status",
 	}
-	return card
+	if _, err = card.Transaction(status); err != nil {
+		return nil, fmt.Errorf("error querying Notecard status: %v", err)
+	}
+	return card, nil
 }
 
 func main() {
-	var i interface{} = os.Getenv("NOTECARD_TRANSPORT")
-	transport, err := i.(Transport)
-
-	if !err {
-		fmt.Printf("Error getting transport protocol, defaulting to I2C...\n")
+	if transport == "" {
+		log.Printf("transport protocol not provided, defaulting to I2C...")
 		transport = I2C
 	}
 
-	card = setupNotecard(transport)
+	card, err := setupNotecard(transport)
+	if err != nil {
+		log.Printf("while setting up notecard: %v", err)
+	}
 	defer card.Close()
```

Comment:
we don't really care given our lifecycle, but this is going to fail (panic even), if the setup above failed (because card will be nil).